Tutorials (#2120)
grafana/agent: docs/flow/tutorials/chaining.md

---
aliases:
- /docs/agent/latest/flow/tutorials/chaining
title: Chaining Prometheus components
weight: 400
---

This tutorial shows how to use [multiple-inputs.flow](../assets/flow_configs/multiple-inputs.flow) to send data to several different locations. This tutorial uses the same base as [Filtering metrics]({{< relref "filtering-metrics.md">}}). 

A new concept introduced in Flow is chaining components together in a composable pipeline. This promotes the reusability of components while offering flexibility. 

# Prerequisites

* [Docker](https://www.docker.com/products/docker-desktop)
* Clone the [Agent Repository](https://github.com/grafana/agent) 
    * `git clone git@github.com:grafana/agent.git`

## Example

To run the example, execute `CONFIG_FILE=multiple-inputs.flow docker-compose -f ./assets/docker-compose.yaml up` from the tutorials directory. 

Allow the Agent to run for two minutes, then go to [Grafana](http://localhost:3000/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Mimir%22,%7B%22refId%22:%22A%22,%22instant%22:true,%22range%22:true,%22exemplar%22:true,%22expr%22:%22agent_build_info%7B%7D%22%7D%5D).

![](./assets/multiple.png)

> There are four series&mdash;two scrapers each sending metrics to both filters, so you end up with four series. They differ in `job` and `cool_label/not_cool_label`.

Reviewer: This sentence reads a little strangely. What about something like this?

> There are two scrapes each sending two metrics to both filters, so you end up with a total of four metrics.

Author: I think the suggestion on the sentence feels better. I think we should lean into/educate on metric vs series but this might not be the place to do it.

---
aliases:
- /docs/agent/latest/flow/tutorials/chaining
title: Chaining components
weight: 400
---

The goal of this tutorial is to show the ability to have data go to several different locations using [multiple-inputs.flow](../assets/flow_configs/multiple-inputs.flow). This uses the same base as [filtering metrics]({{< relref "filtering-metrics.md">}}).

A new concept in Flow is chaining components together. 

# Prerequisites

* [Docker](https://www.docker.com/products/docker-desktop)
* Clone the [Agent Repository](https://github.com/grafana/agent) `git clone git@github.com:grafana/agent.git`

## Example

> To run the example execute `docker-compose -f ./assets/multiple-inputs.yaml up` from the tutorials directory. 

Reviewer 2: nit: Doesn't make much difference, but the three docker-compose files are pretty similar, right? How about adding just one and passing the appropriate configuration file as an env var? The instruction would look like

```
CONFIG_FILE=multiple-inputs.flow docker-compose -f ./assets/agent-flow.yaml up
```

Author: Sounds solid, let me give it a try.

Author: Worked great!

Reviewer 2: Nice! I'll keep this in mind, we might be able to reuse this elsewhere too!

---
aliases:
- /docs/agent/latest/flow/tutorials/chaining

> title: Chaining components

Reviewer: I'd still like this to have a Prometheus-related title.

Author: Changed name to "Chaining Prometheus components".

weight: 400
---

> The goal of this tutorial is to show how to use [multiple-inputs.flow](../assets/flow_configs/multiple-inputs.flow) to send data to several different locations. This tutorial uses the same base as [Filtering metrics]({{< relref "filtering-metrics.md">}}). 

Reviewer: ```suggestion
This tutorial shows how to use [multiple-inputs.flow](../assets/flow_configs/multiple-inputs.flow) to send data to several different locations. This tutorial uses the same base as [Filtering metrics]({{< relref "filtering-metrics.md">}}). 
```

A new concept introduced in Flow is chaining components together in a composable pipeline. This promotes the reusability of components while offering flexibility. 

# Prerequisites

* [Docker](https://www.docker.com/products/docker-desktop)
* Clone the [Agent Repository](https://github.com/grafana/agent) 
    * `git clone git@github.com:grafana/agent.git`

## Example

> To run the example execute `CONFIG_FILE=multiple-inputs.flow docker-compose -f ./assets/docker-compose.yaml up` from the tutorials directory. 

Reviewer 3: ```suggestion
To run the example, execute `CONFIG_FILE=multiple-inputs.flow docker-compose -f ./assets/docker-compose.yaml up` from the tutorials directory. 
```

> Allow the Agent to run for two minutes then go to [Grafana](http://localhost:3000/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Mimir%22,%7B%22refId%22:%22A%22,%22instant%22:true,%22range%22:true,%22exemplar%22:true,%22expr%22:%22agent_build_info%7B%7D%22%7D%5D).

Reviewer 3: ```suggestion
Allow the Agent to run for two minutes, then go to [Grafana](http://localhost:3000/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Mimir%22,%7B%22refId%22:%22A%22,%22instant%22:true,%22range%22:true,%22exemplar%22:true,%22expr%22:%22agent_build_info%7B%7D%22%7D%5D).
```

![](./assets/multiple.png)

> There are four series&mdash;two scrapers each sending metrics to both filters, so you end up with 4 series. They differ in `job` and `cool_label/not_cool_label`.

Reviewer 3: ```suggestion
There are four series&mdash;two scrapers each sending metrics to both filters, so you end up with four series. They differ in `job` and `cool_label/not_cool_label`.
```
